base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,370 @@
+/*
+ * PROJECT:     ReactOS Tasklist Command
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Displays a list of currently running processes on the computer.
+ * COPYRIGHT:   Copyright 2020 He Yang (1160386205@qq.com)
+ */
+
+#include "tasklist.h"
+
+
+// the strings in opList are the command-line options.
+// should always correspond with the defines below, in sequence (except OP_PARAM_INVALID)
+static PCWSTR opList[] = { L"?", L"nh" };
+
+#define OP_PARAM_INVALID    -1
+#define OP_PARAM_HELP       0
+#define OP_PARAM_NOHEADER    1
+
+// the max string length PrintResString can handle
+#define RES_STR_MAXLEN 64
+
+VOID PrintSplitLine(INT Length)
+{
+    for (INT i = 0; i < Length; i++)
+    {
+        ConPrintf(StdOut, L"=");
+    }
+}
+
+// Print spaces
+BOOL PrintSpace(INT SpaceNum)
+{
+    for (; SpaceNum; SpaceNum--)
+    {
+        ConPrintf(StdOut, L" ");
+    }
+    return TRUE;
+}
+
+// Print a string.
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+INT PrintString(LPCWSTR String, INT MaxWidth, BOOL bAlignLeft)
+{
+    return ConPrintf(StdOut, bAlignLeft ? L"%-*.*ls" : L"%*.*ls", MaxWidth, MaxWidth, String);
+}
+
+// Print a string from resource
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+// The string will be truncated if it's longer than RES_STR_MAXLEN
+INT PrintResString(HINSTANCE hInstance, UINT uid, INT MaxWidth, BOOL bAlignLeft)
+{
+    WCHAR StringBuffer[RES_STR_MAXLEN];
+    LoadStringW(hInstance, uid, StringBuffer, _countof(StringBuffer));
+    return PrintString(StringBuffer, MaxWidth, bAlignLeft);
+}
+
+// Print a number, aligned to right.
+// MaxWidth is the width for printing.
+// return FALSE if number length is longer than MaxWidth
+BOOL PrintNum(LONGLONG Number, INT MaxWidth)
+{
+    // calculate the length needed to print.
+    INT PrintLength = 0;
+    LONGLONG Tmp = Number;
+
+    if (Tmp < 0)
+    {
+        Tmp = -Tmp;
+        PrintLength++;
+    }
+
+    do
+    {
+        Tmp /= 10;
+        PrintLength += 1;
+    } while (Tmp);
+
+    if (PrintLength > MaxWidth)
+    {
+        return FALSE;
+    }
+    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
+
+    return TRUE;
+}
```

Comment:
```suggestion
VOID PrintNum(LONGLONG Number, INT MaxWidth)
{
    /* NOTE: The maximum length of the printed number will be 20 characters
     * (including the sign for signed, or with another digit for unsigned). */
    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
}
```
Also as a user I don't want a tool that crashes on me just because it's unable to display long yet valid information.

Reply:
there's a problem with this. It won't truncate the number if it's longer than ```MaxWidth```. should I write some code to truncate the number?

Reply:
If it truncates the number it won't report accurate information as well. Not sure this is wanted.

Reply:
I observed ```tasklist``` on windows under TABLE mode (```/fo table```, also the default format), the column width is big enough that the number is almost impossible to need to be truncated.

But the image name column does truncate the file name if it's too long (while won't in ```list``` and ```csv``` format) 

so I think possibly it's more important to keep it well-formatted under table mode

that's why I think truncate the number is better, does this make sense? 😂

Reply:
While truncating strings can make sense, truncating numbers don't.

Reply:
then fine 😂
after all, it's almost impossible to have these number big enough to need to be truncated